### PR TITLE
Parametrization of PeerId in Reactor API

### PR DESCRIPTION
**Pull request by dr-orlovsky**
Required to support arbitrary remote peer types, operating not just as sockets, but through proxy (which will always have the same address for all peers), like SOCKS5-based proxies (used in Tor, I2P, Nym etc)

### CODE REVIEW

File: net/src/lib.rs [hunk 1]
```diff
@@ -82,13 +83,31 @@ impl<T: fmt::Display> fmt::Display for DisconnectReason<T> {
...
 ///
 /// Network protocols must implement this trait to be drivable by the reactor.
-pub trait Protocol: Iterator<Item = Io<Self::Event, Self::DisconnectReason>> {
+pub trait Protocol<Id: PeerId>:
+    Iterator<Item = Io<Self::Event, Self::DisconnectReason, Id>>
+{
     /// Events emitted by the protocol.
     type Event: fmt::Debug;
-    /// Reason a peer was disconnected.
+    /// Reason a peer was disconnected.std
```
Comment (cloudhead) [inline net/src/lib.rs:110]:
```suggestion
    /// Reason a peer was disconnected.
```

File: net/src/lib.rs [hunk 2]
```diff
@@ -103,22 +122,18 @@ pub trait Protocol: Iterator<Item = Io<Self::Event, Self::DisconnectReason>> {
         // figures of children and girls and voices childish and girlish in the air." -JJ
     }
     /// Received bytes from a peer.
-    fn received_bytes(&mut self, addr: &net::SocketAddr, bytes: &[u8]);
+    fn received_bytes(&mut self, addr: &Id, bytes: &[u8]);
     /// Connection attempt underway.
-    ///
+    ///std
```
Comment (cloudhead) [inline net/src/lib.rs:127]:
```suggestion
    ///
```

**Review by cloudhead — COMMENTED** [review]
Thanks! Only suggestion would be to make `PeerId: Copy`, since `SocketAddr: Copy`. It would avoid the clone() calls. There's also some text that ended up in a couple places that needs to be removed. I've commented on that.

File: net/src/lib.rs (hunk 2 above)
Comment (dr-orlovsky) [reply net/src/lib.rs:127]: Interesting where these `std` are coming from ... Sure will remove them

**Comment by dr-orlovsky** [comment]
No more `std`s out of nowhere here

**Review by vincenzopalazzo — APPROVED** [review]

**Comment by cloudhead** [comment]
Merged #101 instead, thanks.

**Pull request closed without merging**